src/visitors/ast.rs

```diff
@@ -121,3 +121,188 @@ impl From<&Package> for PackageInfo {
         }
     }
 }
+
+pub(crate) const SEMANTIC_TOKEN_KEYWORD: u32 = 0;
+pub(crate) const SEMANTIC_TOKEN_NUMBER: u32 = 1;
+pub(crate) const SEMANTIC_TOKEN_STRING: u32 = 2;
+
+#[derive(Clone, Default)]
+pub struct SemanticTokenVisitor {
+    pub tokens: Vec<lsp::SemanticToken>,
+}
+
+impl<'a> Visitor<'a> for SemanticTokenVisitor {
+    fn visit(&mut self, node: walk::Node<'a>) -> bool {
+        match node {
+            walk::Node::PackageClause(node) => {
+                let location = node.base.location.clone();
```

Comment:
nit: sort of seems like the clone should be unnecessary here if all you're doing is copying number values (line, column) from it.

Might be I've made an assumption here and there's something that won't allow this to work without the clone.

As a quick check, you could try to switch this one case to be `let location = &node.base.location;` and see if rustc complains. Right now, I'm not quite sure what the barrier would be. The fields on `Position` are both u32 and should therefore be `Copy`.

Reply:
There's a part of me that thinks I shouldn't care about `clone` until `clippy` complains or I actually have something performance sensitive but, ultimately, I also think that we should be thinking clearly about data structures that we're using. I think the `clone` came from some experimentation early on. In the end, we're just pulling `u32`s out, so no need to clone anyway. Removed the references.

Reply:
My feeling is, if it's trivial to avoid a clone we should probably go ahead and do it. That way, we'll have more resources on hand to waste on more important and/or less trivially avoidable stuff down the road.